doc/optimization.md

```diff
@@ -0,0 +1,86 @@
+# RaptorJIT optimization guide
+
+The first step to optimize an application is to characterize its most
+significant performance problem. First profile the application with
+`vmprofile` and then interpret the results using this step-by-step
+method:
+
+```lua
+-- RaptorJIT optimization procedure expressed as pseudo-code.
+function optimize ()
+   -- Spending nontrivial time in the interpreter? That's a bug - stop it!
+   if vmprofile.interpreted > 1% then
+      eliminate_interpreter()
+   -- Spend time in the garbage collector? That's a bug too - stop it!
+   elseif vmprofile.garbagecollection > 1% then
+      eliminate_allocation()
+   -- If you are not stuck in interpreter or GC then take a deeper look...
+   else
+      -- Spend time in JIT code? Check how efficient it is.
+      if vmprofile.head + vmprofile.loop > moderate then 
+         optimize_jit()
+      end
+      -- Spend time in the FFI? Check how efficient that is.
+      if vmprofile.ffi > moderate then
+          optimize_ffi()
+      end
+   end
+end
+```
+
+The control flow is significant in this method. For example, if the
+application is spending time in the interpreter then _that_ is the
+problem. You would need to eliminate this interpreter use before it
+makes sense look into other issues like garbage collection (which is
+often a secondary effect of running interpreted.)
+
+## Eliminate interpreter mode
+
+If your application is spending significant time in the interpreter
+then you need to look for _blacklisted_ code.
+
+Code is blacklisted when repeated attempts at JIT compilation all fail.
```

Comment:
i have some notes on that.
 - `trace_abort()` calls `penalty_pc()` on the aborted trace's starting bytecode, but has some conditions on trace properties.
 - `penalty_pc()` increments the corresponding `J->penalty[i].val` by duplicating and adding a small [0..15] random number.
 - if the penalty gets over `PENALTY_MAX` (60,000), it calls `blacklist_pc()` which patches the bytecode.
 - if it's not blacklisted, the `hotcount` slot is also set to the same value, making it less likely to be compiled soon.
 - `innerloopleft()` (only called by `rec_loop_inpterp() when an interpreted loop isn't taken (because of its condition)) can also increment the penalty but doesn't blacklist, nor sets hotcount.

in your example, i guess the trace didn't meet the conditions in `trace_abort()` to be actually penalised